### PR TITLE
Use consistent failure mode for inline arrays that have unsupported element fields due to different reasons.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @jjonescz, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs [hunk 1]
```diff
@@ -19372,49 +19413,15 @@ namespace System
...
-    IL_0031:  callvirt   ""void System.IDisposable.Dispose()""
-    IL_0036:  endfinally
-  }
-  IL_0037:  ret
-}
-");
+            comp.VerifyDiagnostics(
+                // (21,27): error CS9185: foreach statement on an inline array of type 'Buffer4<int>' is not supported
+                //         foreach (var y in x.F)
+                Diagnostic(ErrorCode.ERR_InlineArrayForEachNotSupported, "x.F").WithArguments("Buffer4<int>").WithLocation(21, 27)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs:19419]: >ERR_InlineArrayForEachNotSupported [](http://example.com/codeflow?start=37&length=34)

Why is this case no longer supported?

File: src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs [hunk 2]
```diff
@@ -21294,10 +21350,16 @@ public void UserDefinedIndexer_Warning_05()
 }
 ";
             var comp = CreateCompilation(src, targetFramework: TargetFramework.Net80);
-            CompileAndVerify(comp, expectedOutput: "int").VerifyDiagnostics(
+            comp.VerifyEmitDiagnostics(
+                // (3,26): error CS0021: Cannot apply indexing with [] to an expression of type 'Buffer4'
+                // System.Console.WriteLine(b[0]);
+                Diagnostic(ErrorCode.ERR_BadIndexLHS, "b[0]").WithArguments("Buffer4").WithLocation(3, 26),
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs:21356]: > Were we falling through to a non-inline array case in ```BindElementOrIndexerAccess``` previously?

Yes

File: src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs:19419]: > Why is this case no longer supported?

I do not think it is worth supporting fall back for scenarios with malformed span types. At least from the start. The spec says foreaching over an inline array is like foreaching over a span. The span type is malformed in this case.

> Were we falling through to a non-inline array case in ForEachLoopBinder?

Yes.

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jjonescz, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 3]
```diff
@@ -1825,34 +1825,36 @@ protected override void AfterMembersCompletedChecks(BindingDiagnosticBag diagnos
...
+                    foreach (var slice in GetMembers(WellKnownMemberNames.SliceMethodName).OfType<MethodSymbol>())
+                    {
+                        if (Binder.MethodHasValidSliceSignature(slice))
                         {
-                            if (Binder.MethodHasValidSliceSignature(slice))
-                            {
-                                diagnostics.Add(ErrorCode.WRN_InlineArraySliceNotUsed, slice.TryGetFirstLocation() ?? GetFirstLocation());
-                                break;
-                            }
+                            diagnostics.Add(ErrorCode.WRN_InlineArraySliceNotUsed, slice.TryGetFirstLocation() ?? GetFirstLocation());
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1846]: Btw, should we have similar warnings for user-defined `GetEnumerable()` methods?

For example, the following warns for Slice but not for GetEnumerable even though neither of them is used (and both are used if InlineArray attribute is removed):

```cs
[System.Runtime.CompilerServices.InlineArray(4)]
struct Buffer4
{
    private int _element0;

    int Length => 4;
    string Slice(int i, int j) => "int";

    public IEnumerator<char> GetEnumerator()
    {
        yield return 'e';
        yield return 'n';
        yield return 'u';
    }

    static void Main()
    {
        Buffer4 b = default;
        System.Console.WriteLine(b[..][0]);
        
        foreach (var ch in b)
        {
            System.Console.WriteLine(ch);
        }
    }
}
```

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 4]
```diff
@@ -2251,6 +2251,7 @@ internal enum ErrorCode
         ERR_CollectionBuilderAttributeInvalidMethodName = 9186,
         ERR_CollectionBuilderAttributeMethodNotFound = 9187,
         ERR_CollectionBuilderNoElementType = 9188,
+        ERR_InlineArrayForEachNotSupported = 9189,
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:2254]: Hm, I will have to renumber all my error codes again if you merge first https://github.com/dotnet/roslyn/pull/69230/commits/9ba1fee40d532910893f12f0e3e0b1e7c5cbf8ca#diff-3ac66c25866171b5ebdd5ff2622127096c9b78b7f724d7936b64872faa977894

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1846]: > Btw, should we have similar warnings for user-defined GetEnumerable() methods?

There were no requests for that. Besides, the method can still be called explicitly, but there is no way to do that with an indexer or a conversion operator.

**Pull request merged**